### PR TITLE
WIP: Display user account

### PR DESCRIPTION
**Pull request by vjousse**
Referencing https://github.com/n1k0/tooty/issues/10

### CODE REVIEW

File: src/Mastodon.elm [hunk 1]
```diff
@@ -2,6 +2,7 @@ module Mastodon
     exposing
         ( AccessTokenResult
         , Account
+        , AccountId
```
Comment (n1k0) [inline src/Mastodon.elm]: Hmm, I don't see the real benefit of exposing type aliases like this. In my views they're useful internally to this module, but that may just be me.

File: src/Mastodon.elm [hunk 2]
```diff
@@ -438,6 +444,11 @@ send tagger builder =
     builder |> HttpBuilder.send (toResponse >> tagger)
 
 
+fetchAccount : Client -> AccountId -> Request Account
+fetchAccount client accountId =
+    fetch client ("/api/v1/accounts/" ++ (accountId |> toString)) accountDecoder
```
Comment (n1k0) [inline src/Mastodon.elm]: Nit: Maybe just `(toString accountId)`

File: src/Model.elm [hunk 3]
```diff
@@ -265,5 +276,12 @@ update msg model =
                 Err error ->
                     { model | publicTimeline = [], errors = (errorText error) :: model.errors } ! []
 
+        UserAccount account ->
+            let
+                d =
+                    Debug.log "[UserAccount] Loaded " account
```
Comment (n1k0) [inline src/Model.elm]: Hmm, did you left this on purpose? I'm usually relying on the debuuger to track these instead.

File: src/View.elm [hunk 4]
```diff
@@ -26,25 +27,36 @@ errorsListView model =
...
-        Just (Mastodon.Reblog reblog) ->
-            div [ class "reblog" ]
-                [ p []
-                    [ a [ href status.account.url ] [ text <| "@" ++ status.account.username ]
-                    , text " reblogged"
+    let
+        accountLinkAttributes =
+            [ href status.account.url
+              -- When clicking on a status, we should not let the browser
+              -- redirect to a new. That's why we're preventing the default
```
Comment (n1k0) [inline src/View.elm]: Nit: ... to a new what? 😄

File: src/View.elm [hunk 5]
```diff
@@ -26,25 +27,36 @@ errorsListView model =
...
+              -- behavior here
+            , onWithOptions "click"
+                { stopPropagation = True, preventDefault = True }
```
Comment (n1k0) [inline src/View.elm]: We should move this guy to a local helper:

```elm
onClick_ : msg -> Attribute msg
onClick_ msg =
    Events.onWithOptions
        "click"
        { preventDefault = True, stopPropagation = True }
        (Decode.succeed msg)
```

File: src/Model.elm [hunk 6]
```diff
@@ -265,5 +276,12 @@ update msg model =
...
+            in
+                model ! []
```
Comment (n1k0) [inline src/Model.elm]: To be continued I presume 🙂

File: src/Mastodon.elm (hunk 1 above)
Comment (vjousse) [inline src/Mastodon.elm]: Well I need it in the `Model.elm` file, that's why.

File: src/Mastodon.elm (hunk 2 above)
Comment (vjousse) [inline src/Mastodon.elm]: Sure, I just love this Pizza slice operator |>

File: src/View.elm (hunk 4 above)
Comment (vjousse) [inline src/View.elm]: It is left to the reader's imagination :D

File: src/Mastodon.elm (hunk 1 above)
Comment (n1k0) [inline src/Mastodon.elm]: Not necessarily, just use a normal `String` in `Model.elm`.

**Comment by vjousse** [comment]
tooty_account

File: src/Model.elm [hunk 7]
```diff
@@ -253,6 +257,20 @@ update msg model =
                 Err error ->
                     { model | userTimeline = [], errors = (errorText error) :: model.errors } ! []
 
+        OnLoadUserAccount accountId ->
+            {-
+               @TODO
+               When requesting a user profile, we should load a new "page"
```
Comment (n1k0) [inline src/Model.elm]: What do you mean by new page? If you mean updating the URL in the browser, the `Navigation` module exposes `newUrl` and `replaceUrl` which may do what you're after here. But handling browser history (back and fwd) might be more complicated, so you'll probably want to tackle this later.

File: src/Model.elm [hunk 8]
```diff
@@ -269,5 +287,13 @@ update msg model =
                 Err error ->
                     { model | publicTimeline = [], errors = (errorText error) :: model.errors } ! []
 
+        UserAccount result ->
+            case result of
+                Ok account ->
+                    { model | account = Just account } ! []
+
+                Err error ->
+                    { model | account = Nothing, errors = (errorText error) :: model.errors } ! []
```
Comment (n1k0) [inline src/Model.elm]: I feel bad about us keeping duplicating similar logic for these `case..of` statements. Could you see any way to factor them?

File: src/View.elm [hunk 9]
```diff
@@ -60,7 +43,7 @@ statusContentView : Mastodon.Status -> Html Msg
 statusContentView status =
     case status.spoiler_text of
         "" ->
-            div [ class "status-text" ] <| formatContent status.content
+            div [ class "status-text" ] <| (ViewHelper.formatContent status.content status.mentions)
```
Comment (n1k0) [inline src/View.elm]: Nit: The parenthesis aren't needed here.

File: src/View.elm [hunk 10]
```diff
@@ -69,38 +52,91 @@ statusContentView status =
                     "spoiler" ++ (toString status.id)
             in
                 div [ class "status-text spoiled" ]
-                    [ div [ class "spoiler" ] <| formatContent status.spoiler_text
+                    [ div [ class "spoiler" ] <| (ViewHelper.formatContent status.spoiler_text status.mentions)
```
Comment (n1k0) [inline src/View.elm]: Nit: I realize `spoiler_text` doesn't actually contain HTML, but plain text it seems. No big deal.

File: src/View.elm [hunk 11]
```diff
@@ -232,7 +268,13 @@ homepageView model =
         [ draftView model
         , timelineView model.userTimeline "Home timeline" "home"
         , timelineView model.localTimeline "Local timeline" "th-large"
-        , timelineView model.publicTimeline "Public timeline" "globe"
+        , case model.account of
+            Just account ->
+                -- Todo: Load the user timeline
```
Comment (n1k0) [inline src/View.elm]: Yes indeed :) let's file an issue about that.

File: src/ViewHelper.elm [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+{-| Converts nodes to virtual dom nodes.
+-}
+toVirtualDom : List Mastodon.Mention -> List HtmlParser.Node -> List (Html Msg)
+toVirtualDom mentions nodes =
+    List.map (toVirtualDomEach mentions) nodes
+
+
+addOnClickAttributes : List (Attribute msg) -> List (Attribute msg)
+addOnClickAttributes attrs =
+    attrs
```
Comment (n1k0) [inline src/ViewHelper.elm]: W... Why?

File: src/ViewHelper.elm [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+
+createLinkNode : List ( String, String ) -> List HtmlParser.Node -> List Mastodon.Mention -> Html Msg
+createLinkNode attrs children mentions =
+    let
+        maybeMention =
+            getMentionForLink attrs mentions
+    in
+        case maybeMention of
+            Just mention ->
+                Html.node "a" ((List.map toAttribute attrs) ++ [ onClickWithPreventAndStop (OnLoadUserAccount mention.id) ]) (toVirtualDom mentions children)
```
Comment (n1k0) [inline src/ViewHelper.elm]: Nit: This line would sit better on 2 or maybe 3 lines.

File: src/ViewHelper.elm [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+            (\( name, value ) -> value)
+        |> List.head
+
+
+getMentionForLink : List ( String, String ) -> List Mastodon.Mention -> Maybe Mastodon.Mention
+getMentionForLink attrs mentions =
+    let
+        maybeHref =
+            getHrefLink attrs
+    in
```
Comment (n1k0) [inline src/ViewHelper.elm]: Nit: Maybe the `let..in` block isn't really necessary here.

File: src/Model.elm (hunk 7 above)
Comment (vjousse) [inline src/Model.elm]: Well in fact it's really easy to do. https://github.com/elm-lang/navigation and https://github.com/evancz/url-parser are insanely easy to use. I just love to put @TODO everywhere when I on the subject :-)

File: src/ViewHelper.elm (hunk 12 above)
Comment (vjousse) [inline src/ViewHelper.elm]: Ooops, we need to find the guy how did that 🤓

**Comment by vjousse** [comment]
Closes #10

**Pull request merged**